### DISCUSSION TITLE
Keycloak groups contain duplicate users

### DISCUSSION THREAD

**Discussion by fostermi** [keycloak/keycloak#27098] [Q&A]
After upgrading from the legacy wildfly installation to the Quarkus version of KC last month, using LDAP as our User Federation. We now have a situation where all of our groups contain duplicate users for every single group. This did not happen during the upgrade and initial import of the realm. It just magically happened for no apparent reason. We have KC configured for the "groups" mapper to be LDAP_ONLY and it has been working fine until this issue popped up. If I try to remove all users from a group in KC and then immediately try to sync "LDAP groups to KC", it *seems* to update the groups properly and there is only one entry for each user. However, what actually happens is that it deletes the users from LDAP itself. If you look at a User, their group membership is missing.

This happened in both our Staging and Production environments. In staging I tried to stop KC and delete all the entries in the `user_group_membership` table, started KC and then did a `Sync LDAP groups to KC`. This *seemed* to work in that I only see one entry for each user in all the groups. However, the `user_group_membership` table did not re-populate itself.

Does anyone have any idea what happened and how I can get rid of all the duplicates in KC without wiping out my LDAP server info?

**Comment by riton** [comment]
We've just migrated from a legacy KC `8.x` infrastructure to a new `25.x` infrastructure using an _export to JSON files_ and then _import from JSON files_ . Our realm was successfully imported and our _LDAP_ _User Federation_ was properly restored.
A few hours later, we've realized that our users appeared twice in our KC groups.

We're in the **exact** situation you've experienced and described @fostermi :student: 

We're not exactly sure what is happening...
Did you managed to solve this weird behavior ?

Thanks in advance

Regards

**Reply by fostermi** [reply]
So IIRC, it was as mentioned to stop KC, delete all the entries in the `user_group_membership` table, start KC and then finally `Sync LDAP groups to KC`. After this all of the group information was correct, but that table remained empty. I'm not sure exactly how KC is tracking this if not in that table, unless there is some other mechanism for managing Federation.
